### ISSUE TITLE
Parse error in class.phpmailer.php

### ISSUE THREAD

**Issue by chybaDapi** [PHPMailer/PHPMailer#535]
I have the latest version of the PHPMailer (not modified by me yet) and after run example (from examples/mail.phps) I have the error:
> Parse error: syntax error, unexpected T_FUNCTION in /home/[...]/PHPMailer-master/class.phpmailer.php on line 3040

Code (lines 3028 - 3044):
```php
 protected function clearQueuedAddresses($kind)
    {
        if (version_compare(PHP_VERSION, '5.3.0', '<')) {
            $RecipientsQueue = $this->RecipientsQueue;
            foreach ($RecipientsQueue as $address => $params) {
                if ($params[0] == $kind) {
                    unset($this->RecipientsQueue[$address]);
                }
            }
        } else {
            $this->RecipientsQueue = array_filter(
                $this->RecipientsQueue,
                function ($params) use ($kind) {
                    return $params[0] != $kind;
                });
        }
    }
```
A error is in this: function ($params) use ($kind) {;

What can I do?

**Comment by Synchro** [comment]
`array_filter` is expecting a callable there, so it looks correct. What version of PHP are you running?

**Comment by chybaDapi** [comment]
My version of PHP is 5.2.17

**Comment by Synchro** [comment]
That doesn't make much sense - there is a version check right there that means that code should never be run on a version that old. Why are you doing new development on such an old version of PHP? It's been out of support for nearly 5 years, so it's positively dangerous to deploy anything on it.

**Comment by ivantcholakov** [comment]
The author of this code fragment had the intention of supporting PHP 5.2.x, but code is simply wrong. Also, I don't see a reason for rewriting this simple algorithm twice. I can imagine this quick patch:

```php
    /**
     * Clear queued addresses of given kind.
     * @access protected
     * @param string $kind 'to', 'cc', or 'bcc'
     * @return void
     */
    protected function clearQueuedAddresses($kind)
    {
        //if (version_compare(PHP_VERSION, '5.3.0', '<')) {
            $RecipientsQueue = $this->RecipientsQueue;
            foreach ($RecipientsQueue as $address => $params) {
                if ($params[0] == $kind) {
                    unset($this->RecipientsQueue[$address]);
                }
            }
        //} else {
        //    $this->RecipientsQueue = array_filter(
        //        $this->RecipientsQueue,
        //        function ($params) use ($kind) {
        //            return $params[0] != $kind;
        //        });
        //}
    }
```

**Comment by Synchro** [comment]
Yes, I see the issue - the version check doesn't prevent PHP from trying to compile the code, so version checks like this are OK for things like function names, but no use for language constructs. I'll fix it.

**Comment by Synchro** [comment]
Fixed in HEAD

**Comment by ivantcholakov** [comment]
Thank you. :-)

**Comment by fbonzon** [comment]
I [committed](https://github.com/PHPMailer/PHPMailer/commit/0fa95186090448ead097867b93fd597782190dff) this code 7 days ago – my bad. I learned something today: version checks are good only if also using valid language constructs in older versions we want supported.

I can add that I looked at the [Travis CI report](https://travis-ci.org/PHPMailer/PHPMailer/builds/85403803) of the pull request for this commit. It obviously didn't report any problem… because Travis CI supports PHP versions only down to 5.3. It's time to have PHPMailer minimum requirements in sync with the developer tools, and I was happy to read 5 days ago in #323 that PHPMailer is soon going to require PHP 5.4. This will avoid such problems in the future.

@ivantcholakov: A single `array_filter()` call is faster and more memory friendly that copying the whole array first, followed by a `foreach` loop, thus my initial intention to make the code more efficient for PHP 5.3+. It may be duplicated simple algorithm, but not "wrong" IMHO.
